### PR TITLE
ParamType improvements

### PR DESCRIPTION
**Pull request by namutaka**
ParamType.String,DateTime,LocalDate,LocalTime consider empty string as null.

### CODE REVIEW

**Comment by seratch** [comment]
Thanks a lot. I'll merge this after build success.

File: common/src/main/scala/skinny/ParamType.scala [hunk 1]
```diff
@@ -79,22 +79,24 @@ object ParamType {
...
   case object ByteArray extends AbstractParamType({
-    case v: String if v == null || v.trim.isEmpty => null
+    case v: String if v.trim.isEmpty => null
     case v: String => v.getBytes
     case v: Array[Byte] => v
   })
   case object DateTime extends ParamType {
     def unapply(v: Any): Option[Any] = v match {
       case null => None
+      case v: String if v.trim.isEmpty => Some(null)
```
Comment (gakuzzzz) [inline common/src/main/scala/skinny/ParamType.scala]: Why it is not None ?

File: common/src/main/scala/skinny/ParamType.scala [hunk 2]
```diff
@@ -79,22 +79,24 @@ object ParamType {
     case v: Short => scala.math.BigDecimal(v)
   })
   case object String extends AbstractParamType({
+    case v: String if v.isEmpty => null // don't trim
```
Comment (gakuzzzz) [inline common/src/main/scala/skinny/ParamType.scala]: Is it means that we can not use empty string parameter?

File: common/src/main/scala/skinny/ParamType.scala (hunk 1 above)
Comment (seratch) [inline common/src/main/scala/skinny/ParamType.scala]: Ah, this should be None.

File: common/src/main/scala/skinny/ParamType.scala (hunk 2 above)
Comment (seratch) [inline common/src/main/scala/skinny/ParamType.scala]: Yes, it does. I think this is reasonable enough in real world web apps. But we may miss some use cases.

Comment (gakuzzzz) [inline common/src/main/scala/skinny/ParamType.scala]: Hmm, We often make column `not-null` as like `comment` and accept empty String.
In this case, can we convert None to empty String?

File: common/src/main/scala/skinny/ParamType.scala (hunk 1 above)
Comment (seratch) [inline common/src/main/scala/skinny/ParamType.scala]: http://www.scala-lang.org/api/current/index.html#scala.PartialFunction$

We found that `PartialFunction.condOpt` unexpectedly (for me) returns `Some(null)` in other `ParamType`s. Though this is another issue, let's fix to return `None` in this pull request.

File: common/src/main/scala/skinny/ParamType.scala (hunk 2 above)
Comment (seratch) [inline common/src/main/scala/skinny/ParamType.scala]: I see. Indeed, we should care about such a case. It's not good for existing apps to change this behavior now.

@namutaka How about defining your own `ParamType` in your case?

```scala
case object NonEmptyString extends AbstractParamType({
  case v: String if v.isEmpty => null // don't trim
  ..
})
```

If lots of users need the above case object, we may need to consider put it into framework.

Comment (namutaka) [inline common/src/main/scala/skinny/ParamType.scala]: Okey, I use own ```ParamType``` in my apps.

File: common/src/main/scala/skinny/ParamType.scala (hunk 1 above)
Comment (seratch) [inline common/src/main/scala/skinny/ParamType.scala]: @namutaka Sorry, my mistake. In this case, existing parameter is treated as null. So `Some(null|)` is reasonable. So please remove ParamType.String modification. Others look good to me.

**Comment by seratch** [comment]
Thanks.

**Pull request merged**